File: mac/KhaytCore/Sources/KhaytCore/GeometryKey.swift
import Foundation

/// The key for "the same mesh, however it was packaged" — ported to Swift.
///
/// ── THE FIRST MODULE OF THE PORT, AND WHY IT IS THIS ONE ──────────────────
///
/// It is three numbers joined by punctuation, which `lib/geometry-key.js` says
/// is "exactly the kind of thing two implementations agree on until they do
/// not: a rounding, a separator, an order". That makes it the honest first
/// test of whether a port can hold: if Swift can reproduce this byte for byte
/// over generated input, the method works; if it cannot, better to learn that
/// on seventy lines than on seven hundred.
///
/// Every key here is compared against records the other app wrote, so this is
/// not a re-derivation of what a geometry key OUGHT to be. It is the same
/// answer, spelled the same way, including the parts that look like accidents:
/// `Math.round`'s half-up, and a whole number printed without a `.0`. Both live
/// in `JSSemantics` and both are why this is a port rather than a rewrite.
///
/// `test/…`'s JavaScript remains the specification. `GeometryKeyParityTests`
/// runs both over generated geometry and fails on the first disagreement.
public enum GeometryKey {

    /// Which reader measured a record, and the number that goes up when a
    /// reader fault is fixed.
    ///
    /// Reader 2: one plate's size rather than every plate boxed together,
    /// every component placed, roots over 8 MB read, zip64 containers opened.
    public static let reader = 2

    /// `round(v, dp)` — nil where the module returns null.
    public static func round(_ value: JSONValue?, _ dp: Int) -> Double? {
        let n = JSSemantics.number(value)
        guard n.isFinite else { return nil }
        let f = pow(10.0, Double(dp))
        return JSSemantics.round(n * f) / f
    }

    /// The key, or nil for geometry with no substance.
    ///
    /// Nil rather than a key, so an unparsed or empty model never acquires an
    /// identity that another empty one would share.
    public static func key(of geometry: JSONValue?) -> String? {
        guard case .object(let g)? = geometry else {
            // `geometry || {}` in the module: anything that is not an object
            // reads as an empty one, and an empty one has no triangles.
            return nil
        }
        let tris = JSSemantics.number(g["triangleCount"])
        let volume = JSSemantics.number(g["volumeMm3"])
        guard tris.isFinite, tris > 0 else { return nil }
        guard volume.isFinite, volume > 0 else { return nil }

        var box: [String: JSONValue] = [:]
        if case .object(let b)? = g["bbox"] { box = b }
        guard let x = round(box["x"], 2), let y = round(box["y"], 2),
              let z = round(box["z"], 2), let vol = round(.number(volume), 2)
        else { return nil }

        let dims = [x, y, z].map(JSSemantics.string).joined(separator: "x")
        return "\(JSSemantics.string(tris)):\(JSSemantics.string(vol)):\(dims)"
    }

    /// Whether a record is due to be measured again.
    public static func needsRemeasure(_ record: JSONValue?) -> Bool {
        var value: JSONValue?
        if case .object(let r)? = record { value = r["geometryReader"] }
        let n = JSSemantics.number(value)
        return !(n.isFinite && n >= Double(reader))
    }
}

File: lib/geometry-key.js
'use strict';
(function (global) {

/**
 * The key for "the same mesh, however it was packaged".
 *
 * Split out of `lib/model-identity.js`, which cannot be shared with the Mac
 * app: its `contentHash` falls back to `require('crypto')` and JavaScriptCore
 * has no `require`. The repo's drift guards refuse a bundled module that names
 * Node at all — rightly, since a guarded require is still a require somebody
 * will later unguard — so the pure half lives here and the whole half keeps
 * re-exporting it.
 *
 * WHY IT IS SHARED AT ALL. It is three numbers joined by punctuation, which is
 * exactly the kind of thing two implementations agree on until they do not: a
 * rounding, a separator, an order. These keys are compared against records the
 * other app wrote, so the format has to have one author.
 *
 * Triangle count is exact on purpose: a re-container preserves it, and a re-mesh
 * does not — and a re-mesh IS a different model as far as a print shop is
 * concerned, because it will slice differently.
 *
 * Returns null for geometry with no substance, so an unparsed or empty model
 * never acquires an identity that another empty one would share.
 *
 * ── WHAT IT CANNOT TELL APART: A PART AND ITS MIRROR ─────────────────────
 *
 * A mirror preserves the triangle count, the bounding box and the volume,
 * which is the whole key. Measured on a real shop's library: 28 sets of
 * "duplicates" by this key, and almost every one a mirrored pair — a left arm
 * against a right arm. Byte-identical duplicates there: zero.
 *
 * The signed volume looks like the missing information and is not. Both files
 * in that pair measure POSITIVE and agree to three decimal places, because the
 * tool that mirrored them flipped the winding to keep the normals outward.
 *
 * So a match here means THE SAME SIZE AND SHAPE, never "the same model".
 * Anything that offers to delete one of a matching pair must say which it is
 * offering, and must not call it a duplicate. See
 * `test/geometry-key-mirror.test.js`.
 */
function round(v, dp) {
  const n = Number(v);
  if (!Number.isFinite(n)) return null;
  const f = Math.pow(10, dp);
  return Math.round(n * f) / f;
}

function geometryKey(geometry) {
  const g = geometry || {};
  const tris = Number(g.triangleCount);
  const vol = Number(g.volumeMm3);
  if (!Number.isFinite(tris) || tris <= 0) return null;
  if (!Number.isFinite(vol) || vol <= 0) return null;
  const bbox = g.bbox || {};
  const dims = [round(bbox.x, 2), round(bbox.y, 2), round(bbox.z, 2)];
  if (dims.some((d) => d === null)) return null;
  return `${tris}:${round(vol, 2)}:${dims.join('x')}`;
}

/**
 * Which reader wrote a record's key, and whether it is due to be read again.
 *
 * A key is written once, on import, and nothing re-reads a file that has one —
 * right for the ordinary case, and exactly what strands a book when a reader
 * fault is fixed: the number the shop was shown stays wrong on every record
 * the old reader wrote. So each record names the reader that measured it, and
 * this number goes up when a fault is fixed. A record below it — or without
 * one, written before the rule existed — is measured again by whichever app
 * next opens the book with the file in reach, and only the keys that differ
 * change. Reader 2: one plate's size rather than every plate boxed together,
 * every component placed, roots over 8 MB read, zip64 containers opened.
 */
const READER = 2;

function needsRemeasure(record) {
  const r = Number(record && record.geometryReader);
  return !(Number.isFinite(r) && r >= READER);
}

const api = { geometryKey, round, READER, needsRemeasure };
if (typeof module !== 'undefined' && module.exports) module.exports = api;
global.KhaytGeometryKey = api;

})(typeof globalThis !== 'undefined' ? globalThis : this);
